src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns_Constant.cs

```diff
@@ -0,0 +1,159 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using System;
+using Microsoft.CodeAnalysis.CodeGen;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class LocalRewriter
+    {
+        private BoundExpression LowerConstantPattern(BoundConstantPattern pattern, BoundExpression input)
+        {
+            return CompareWithConstant(input, pattern.Value);
+        }
+
+        private BoundExpression CompareWithConstant(BoundExpression input, BoundExpression boundConstant)
+        {
+            // PROTOTYPE(patterns): We can generate excellent code when the input is a constant,
+            // PROTOTYPE(patterns): but we don't handle that today.
+            // PROTOTYPE(patterns): Similarly, we can generate better code when the input is statically
+            // PROTOTYPE(patterns): an integral type, but we currently treat all inputs the same as `object`.
+            var value = boundConstant.ConstantValue;
+            if (!value.IsIntegral)
+            {
+                // We use object.Equals for char, string, float, double, decimal, DateTime, and null
+                return _factory.StaticCall(
+                    _factory.SpecialType(SpecialType.System_Object),
+                    // PROTOTYPE(patterns): Add a special member for `public static bool object.Equals(object, object)`
+                    "Equals",
+                    _factory.Convert(_factory.SpecialType(SpecialType.System_Object), input),
+                    _factory.Convert(_factory.SpecialType(SpecialType.System_Object), boundConstant)
+                    );
+            }
+
+            // We use one of three helper functions from PrivateImplementationDetails, depending on the value
+            // of the constant. The helper function takes the input and produces an integral value that we
+            // are to compare against the input. Then we generate code like so:
+            //     SelectedHelperMethod((object)input, out var temp) && temp == Constant
```

Comment:
Would it be possible to incorporate the equality test into the helper? It would be much simpler if we could call the helper ```SelectedHelperMethod((object)input, constant)``` and get the final true or false back.